[로또 미션] 김명지 미션 제출합니다.

Starlight258 opened the pull request:
안녕하세요 진영님 !
로또 2단계 미션을 완료했습니다 😁

## 궁금한 점
- 댓글에 남겼습니다!

✅ 구현된 코드에 대한 설명은 `README.md`에서 보실 수 있습니다 ! 

리뷰 부탁드립니다 !! 감사합니다~ 🙇‍♀️

Starlight258 committed:
feat: Lotto 도메인 구현

- Lotto: 1부터 45 중 무작위 6개 숫자를 포함하는 리스트를 가지는 일급 컬렉션

Starlight258 committed:
feat: InputView, OutputView 구현

Starlight258 committed:
feat: Controller, Application 구현

Starlight258 committed:
feat: Lottos 생성 및 Lotto 출력 로직 캡슐화

- List<Lotto>를 불변으로 만들어 보호하기 위해 Lottos 일급 컬렉션을 생성한다.
- outputView에서 Lottos 정보를 출력하는 두 메서드를 하나로 합친다.

Starlight258 committed:
refactor: Lotto 리팩토링

- 매직넘버 제거
- toList -> getLotto로 변경하여 뜻을 명확하게 한다.
(toList의 경우 Lotto를 List로 만드는 것인지, Lotto 내부 리스트를 가리키고 있는지 명확하지 않음)

Starlight258 committed:
fix: 로또 생성시 45가 포함되지 않는 문제 수정

Starlight258 committed:
test: Lotto 생성 테스트 작성

Starlight258 committed:
refactor: assertAll 사용하여 모든 검증 수행

Starlight258 committed:
rename: lotto 패키지 추가

- Lotto와 Lottos를 lotto 패키지로 묶는다.

Starlight258 committed:
refactor: Lottos 리팩토링

- toList -> getLottos로 메서드명을 변경하여 작업이 명확하게 드러나도록 한다.

Starlight258 committed:
fix: 더 큰 범위의 예외로 수정

- 가격이 숫자로 들어오지 않을 경우와 0 미만일 경우 다시 입력받을 수 있도록 더 큰 범위의 예외로 수정

Starlight258 committed:
refactor: 상수 접근제어자 private으로 변경

- 클래스 내부에서만 사용하므로 상수는 private 접근제어자로 둔다.
- 만약 public 상수를 선언할 경우에는 enum으로 정의하는 것이 타입 안정성을 제공하여 좋다.

Starlight258 committed:
test: 검증값을 상수값이 아닌 값으로 표현하도록 수정

- 상수나 변수로 검증값을 둘 경우 테스트시 필요없는 의존성이 생겨 지양한다.
- 검증은 최대한 간단하게 수행한다. 변수나 상수를 이용하여 값 검증시 해당 값을 확인하기 위해 코드 위쪽까지 거슬러 올라가야한다.

Starlight258 committed:
refactor: Lotto 리팩토링

- 상수 리스트 생성시에만 사용되는 작업은 static 블록으로 이동한다.
- 수동 로또 생성을 위한 정적 팩토리 메서드를 제공한다.
- 불변 리스트 + stream 조합을 사용하여 불변성을 지키며 결과 리스트를 얻는다.
- Collections.shuffle은 원본 리스트를 변경하므로 stream 연산에서 무작위 정수를 사용해 로또를 얻도록 수정한다.

Starlight258 committed:
refactor: stream을 이용한 Lotto 리스트 생성

- Stream을 이용해 Lotto 리스트를 생성하고 toList()를 호출하여 불변 리스트를 반환한다.
- 당장 사용하지 않는 Lottos는 제거한다.

Starlight258 committed:
refactor: 불필요한 정적 팩토리 메서드 제거

- 단순 객체 생성은 생성자로 충분하기 때문에 정적 팩토리 메서드를 제거한다.

Starlight258 committed:
feat: 로또 당첨 1차 구현

Starlight258 committed:
feat: LottoNumber 값 객체 생성

- 원시값을 포장하고 로또 숫자에 대한 검증 책임을 할당한다.

Starlight258 committed:
feat: Lottos 일급 컬렉션 생성

- 로또 리스트를 일급 컬렉션으로 관리하여 로또 리스트에 대한 로직을 한 곳에 모은다.

Starlight258 committed:
refactor: Lotto, Price 리팩토링

- 로또 하나의 가격을 로또 클래스에 둠으로써, 가격은 로또와 관련없이 가격 자체의 로직만 존재하도록 변경한다.
 - 가격 VO에 대한 연산 메서드를 정의한다.

Starlight258 committed:
refactor: 컨트롤러 리팩토링

- 메서드 위치 변경
- 바뀐 코드에 맞춰 메서드명 수정

Starlight258 committed:
feat: LottoResult 일급 컬렉션 생성

- 당첨 통계 로직을 LottoResult에 모은다.

Starlight258 committed:
refactor: 메서드 순서 변경

Starlight258 committed:
refactor: record를 class로 변경

- record는 필드가 private이 아닌 public으로 선언되므로 외부에서 내부 요소에 접근이 가능하다.
- 클래스로 변경하고 Private 필드로 선언하여 직접 접근을 막고 getter 메서드를 통해서만 접근하도록 한다.

Starlight258 committed:
fix: 로또 비교 연산 에러 수정

- equals, hashcode 재정의하여 로또 비교 연산을 정상적으로 수행한다.

Starlight258 committed:
refactor: 코드 리팩토링

- 상수 제거
- 필요없는 변수 제거

Starlight258 committed:
test: Lotto 테스트 작성

Starlight258 committed:
test: 로또 번호 테스트 작성

Starlight258 committed:
test: 로또 묶음 테스트 작성

Starlight258 committed:
test: Price 테스트 작성

Starlight258 committed:
test: LottoResult 테스트 작성

Starlight258 committed:
refactor: 테스트 코드 리팩토링

- 로또 숫자 -> 로또 번호로 명칭 통일
- 예외 메세지 명확하게 수정

Starlight258 committed:
refactor: 코드 리팩토링

- Price를 record에서 class로 변경하여 값에 직접 접근하지 못하도록 막는다.
- 상수를 제거한다.

Starlight258 committed:
fix: 수익률 계산 에러 해결

- 수익률에 100을 곱하여 올바른 값이 나오도록 수정한다.

Starlight258 committed:
refactor: 출력 메서드 분리하여 depth 줄이기

Starlight258 committed:
docs: README 업데이트

Starlight258 commented on src/main/java/org/duckstudy/model/lotto/Lottos.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    public static Lottos generateLottosByPrice(Price price) {
+        int lottoCount = Lotto.calculateLottoCount(price);
```
Lottos 클래스에서 다른 클래스의 calculateLottoCount 메서드를 직접 호출하고 있는데, 이렇게 구현하는 것이 좋은 방법인지 궁금합니다. 클래스 간의 결합도를 낮추기 위한 좋은 방법이 있을까요?

jinyoungchoi95 replied:
어짜피 `Lotto`의 일급컬랙션이기 때문에 `Lotto`에 의존되어있는 구조자체가 이상하지는 않아요.

가격에 대한 책임이라고 한다면 그냥 로또의 가격은 `Price`가 가져도 되는 것 아닐까싶어요. 로또라는 네이밍으로 인해서 `Lotto`가 가격을 가져야한다기보다는 적절한 책임이 어디에 가야할까를 고민해보면 가격에 대한 책임은 모두 `Price`가 가지는 게 맞아보이네요.

Starlight258 replied:
그렇군요 !! 가격에 대한 책임은 모두 `Price`에 있는게 더 적절해 보이네요 !

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/lotto/LottoNumber.java:
```diff
@@ -0,0 +1,44 @@
+
+import java.util.Objects;
+
+public class LottoNumber {
```
👍

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/lotto/LottoNumber.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    private void validateNumber(int number) {
+        if (number < START_INCLUSIVE_NUMBER || number >= END_EXCLUSIVE_NUMBER) {
+            throw new IllegalArgumentException("로또 번호는 1 이상 45 이하의 숫자여야 합니다.");
+        }
```
1이상 45 이하라면 `END_EXCLUSIVE_NUMBER`는 45를 두어 로직을 만드는게 어떨까싶어요.

Starlight258 replied:
그렇게 하면 요구사항이 코드에 명확하게 드러나겠네요! 수정하겠습니다 !

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/lotto/LottoNumber.java:
```diff
@@ -0,0 +1,44 @@
+
+    private final int value;
+
+    public LottoNumber(int number) {
```
1~45만을 사용한다면 값을 캐싱해보는 방법도 있어요.

`Integer`나 `Long`의 구현체를 확인해보시면 도움이 될 것 같아요.

Starlight258 replied:
`Integer` 클래스 구현체를 확인해봤는데, 기본적으로 -128부터 127까지 범위의 정수값에 대해서는 미리 생성하여 `IntegerCache`라는 캐시에 저장해두네요! 
`LottoNumber`에서도 1 ~ 45사이의 값의 객체들만 생성되기 때문에, 캐시에 저장해서 재사용할 수 있도록 하는게 좋겠어요~ 
감사합니다 !!

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/Price.java:
```diff
@@ -15,7 +19,40 @@ private void validatePrice(int price) {
+        return new Price(value * times);
+    }
+
+    public Price dividedBy(int divisor) {
+        return new Price(value / divisor);
+    }
+
+    public double dividedBy(Price divisor) {
+        return (double) value / divisor.getValue();
+    }
```
0으로 나뉘면 어떻게 되나요?

Starlight258 replied:
헉 0원으로 나눌 경우를 미처 생각하지 못했네요..! 테스트에도 추가하겠습니다 !

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/lotto/Lotto.java:
```diff
@@ -1,41 +1,76 @@
-        List<Integer> result = NUMBERS.stream()
-                .sorted(Comparator.comparingInt(i -> ThreadLocalRandom.current().nextInt()))
+        List<LottoNumber> result = NUMBERS.stream()
+                .collect(collectingAndThen(toList(),
+                        list -> {
+                            shuffle(list);
+                            return list.stream();
+                        }))
```
함수가 좀 복잡한 것 같아요. 자바 8에서 함수형프로그래밍이 나온 이유를 생각해보면 행위를 잘 표현하기 위해서 stream이 나온 것인데, 로직이 좀 복잡하지 않나요? 메소드로 분리하거나 할수도 있을 것 같아요.

Starlight258 replied:
맞습니다 ! 리팩토링하겠습니다 !!

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/lotto/Lotto.java:
```diff
@@ -1,41 +1,76 @@
     }
 
-    public List<Integer> getLotto() {
+    public static int calculateLottoCount(Price price) {
+        return price.dividedBy(PER_PRICE).getValue();
+    }
```
`dividedBy`의 결과가 `getValue()`로만 사용된다면 반환값을 `Price`가 아닌 사용되는 int를 바로 반환하게끔 하여도 되지 않을까요?

Starlight258 replied:
맞습니다! `divideBy`에서 int 결과값만 사용하기 때문에 굳이 `Price`로 반환해서 `getValue()`를 호출할 필요가 없을 것 같아요.
감사합니다~

jinyoungchoi95 commented on src/test/java/org/duckstudy/model/lotto/LottosTest.java:
```diff
@@ -0,0 +1,63 @@
+        @Test
+        @DisplayName("당첨 로또 번호를 입력하면 로또 묶음의 당첨 결과를 계산한다")
+        void calculateWinningResultWhenInputWinningLotto() {
+            Lotto winningLotto = Stream.of(1, 2, 3, 4, 5, 6)
+                    .map(LottoNumber::new)
+                    .collect(collectingAndThen(toList(), Lotto::new));
+
+            List<Lotto> lottos = Stream.of(
+                            Stream.of(1, 2, 3, 4, 5, 6),
+                            Stream.of(1, 2, 3, 7, 8, 9),
+                            Stream.of(1, 2, 3, 10, 11, 12)
+                    )
+                    .map(stream -> stream
+                            .map(LottoNumber::new)
+                            .collect(collectingAndThen(toList(), Lotto::new)))
+                    .toList();
```
테스트를 위해 생성하는 로직이 너무 복잡해요. 아래와 같은 방법을 고려해보세요.

1. 생성자에서 추가로 제공하면 되는 것 아닌가
```java
public Lotto(LottoNumber... lottoNumbers) {
    this(List.of(lottoNumbers));
}
```
2. 별도의 생성하는 메소드로 분리한다.

2가 일반적으로 생각나는 방법일텐데 1에서 생성자를 추가하는 방법을 고민해보시면 좋을 것 같아요. 테스트를 위한 생성자를 제가 이전 피드백에서 드렸는지 잘 기억이 나지 않는데, 테스트를 위한 생성자와 테스트를 위한 메소드에 대한 고민을 한번 해보시면 좋을 것 같아요.

Starlight258 replied:
생각해보면 좋은 주제네요 감사합니다 😊

테스트를 위한 메서드를 생성하는 것이 항상 좋은 선택은 아니지만 테스트를 용이하게 하기 위해 적당히 선택하면 좋을 것 같아요!
언급해주신 코드에서도 구현된 메서드만 가지고 테스트를 하려다보니 중복되는 코드도 많고 굉장히 복잡해졌네요..!
테스트를 위한 코드를 주석이나 메서드 명으로 프로덕션 코드와 구분만 잘 짓는다면, 상황에 따라 유연하게 선택하면 좋을 것 같아요.

```java
public static Lotto createLottoForTest(int... values) {
      return new Lotto(IntStream.of(values)
              .mapToObj(LottoNumber::valueOf)
              .toList());
  }
```
메서드에 테스트를 위한 용도를 알리고 싶어서 정적 팩토리 메서드로 구현하려고하는데 어떻게 생각하시나요?
굳이 정적 팩토리 메서드를 생성할 필요 없이 생성자 + 주석으로 충분할까요?

감사합니다 :)

jinyoungchoi95 replied:
아뇨 프로덕션에 Test라는 것을 명시하라라는 뜻은 아니었어요. 별도 생성하는 메소드는 테스트에 정의하는 걸 말씀드린 것이었어요.

테스트를 위한 생성자라기보다는 다양한 형태의 생성자는 객체 생성의 유연성을 높여주기 때문에 권유드린 것이었고, 이를 정적 팩토리 메소드로 표현하는 것은 달라요. 정적 팩토리 메소드와 생성자 그리고 메소드 이 3개는 엄연히 다른 존재들이에요.

테스트를 위한 메소드를 만들지 말라라는 것은 테스트를 위해서 객체의 행위가 추가되지 말라는 뜻이고, 생성자는 안정적인 객체 생성만 보장한다면 유연한 객체 생성을 도와주기 때문에 권장하기도 하는 것이에요. 그리고 정적 팩토리 메소드는 사실상 static이기 때문에 객체지향과 멀어지는 지름길이 될 수도 있고요.

Starlight258 replied:
아 그렇군요 제가 잘못 이해했네요 !! 감사합니다 !

테스트를 위한 별도의 메서드는 테스트 코드 내에서 정의하고, 프로덕션 코드에서는 테스트를 위한 코드를 추가하지 않아야겠네요!
그리고 생성자를 사용해서 객체 생성을 유연하게 제공할 수 있고,
 정적 팩토리 메서드는 클래스 수준에서 동작하기 때문에 객체 지향과 멀어지겠네요!
 
 반영해서 수정하겠습니다. 답변 감사합니다 😊

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/lotto/LottoResult.java:
```diff
@@ -0,0 +1,37 @@
+        this.result = Collections.unmodifiableMap(result);
+    }
+
+    public int getMatchingCount(int count) {
+        return result.getOrDefault(count, ZERO_COUNT);
+    }
```
private 메소드여야겠네요.

Starlight258 replied:
앗 감사합니다!

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/lotto/LottoResult.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    public double calculateProfitRate(Price price) {
+        Price profit = new Price(INITIAL_PRICE);
```
0원을 가진 `Price`는 `Price`에서 정적팩토리메소드로 제공하는 것이 더 적절한 책임이 아닐까요?

Starlight258 replied:
맞습니다! Price에서 0원 객체 생성 책임을 가지는게 캡슐화 관점에서 좋을 것 같아요~
Price에서 생성하면 LottoResult 클래스의` INITIAL_PRICE `상수도 함께 제거할 수 있겠네요! 😀

jinyoungchoi95 reviewed (changes requested):
안녕하세요 명지님 😃

몇가지 코멘트남겨두었으니 확인부탁드려요.

감사합니다🙇‍♂️

Starlight258 committed:
refactor: 가격 책임 Price로 이동

- 로또 가격과 당첨 금액을 Price 클래스로 이동한다.

Starlight258 committed:
refactor: 요구사항이 드러나는 상수로 변경

- 요구사항이 1~45사이의 숫자를 생성하는 것이므로 상한값을 45로 설정한다.

Starlight258 committed:
refactor: LottoNumber에 캐싱 적용

- LottoNumber는 1부터 45사이의 값만 가지므로 캐시에 저장해두고 필요할때마다 재사용하여 불필요한 객체 생성을 줄인다.

Starlight258 committed:
fix: 0원으로 나눌 경우 예외 처리

- 0원으로 나눌 경우 예외를 발생한다.
- 관련 테스트를 추가한다.

Starlight258 committed:
refactor: stream 연산 작업 메서드로 분리

- 메서드로 분리하여 가독성을 향상한다.

Starlight258 committed:
fix: cache 배열 크기 조정

- cache에 46 숫자가 포함되지 않도록 제거한다.

Starlight258 committed:
refactor: 상수명 변경 및 필요없는 메서드 호출 제거

- 요구사항이 드러나도록 상수명을 변경한다.
- 메서드의 결과는 getValue()를 통해서만 사용하기 때문에 int값을 직접 반환하도록 하여 getValue() 호출을 방지한다.

Starlight258 committed:
refactor: 테스트를 위한 메서드를 도입하여 테스트 단순화

- 기존 코드는 너무 복잡하여 가독성이 떨어진다. 테스트를 위한 메서드를 도입한다.
- 테스트를 위한 메서드는 프로덕션 코드와 분리하기 위해 메서드명에 ForTest를 붙인다.

Starlight258 committed:
refactor: Price에서 0원 객체를 생성하는 정적 팩토리 메서드 추가

- 0원 Price 객체 생성을 정적 팩토리 메서드로 제공하여 값 객체 생성 책임을 LottoResult로부터 제거하여 캡슐화한다.

Starlight258 commented:
@jinyoungchoi95 
리뷰해주신 사항들 덕분에 많이 배웠습니다 🙇‍♀️
감사합니다 ! 😁

Starlight258 committed:
refactor: 예외 메세지에 String.format 사용

- 검증값은 바뀔 수 있으므로 그에 따라 예외 메세지도 바뀌도록 수정한다.

Starlight258 committed:
test: 테스트 코드 리팩토링

- 테스트 메서드를 사용하여 복잡한 로직을 대체한다.
- 빈줄을 제거한다.

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/Price.java:
```diff
@@ -15,7 +25,56 @@ private void validatePrice(int price) {
+        return new Price(value * times);
+    }
+
+    public int divideBy(int divisor) {
+        checkIfZero(divisor);
+        return value / divisor;
+    }
```
해당 메소드는 여기서만쓰이니 private으로 변경해주어야할 것 같아요.

Starlight258 replied:
맞습니다..! 감사합니다 !

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/lotto/LottoNumber.java:
```diff
@@ -0,0 +1,56 @@
+
+    private static final int START_INCLUSIVE_NUMBER = 1;
+    private static final int END_INCLUSIVE_NUMBER = 45;
+    private static final LottoNumber[] cache = new LottoNumber[END_INCLUSIVE_NUMBER];
+
+    static {
+        for (int i = 0; i < END_INCLUSIVE_NUMBER; i++) {
+            cache[i] = new LottoNumber(START_INCLUSIVE_NUMBER + i);
+        }
+    }
```
IntStream을 적절하게 활용하여 `List<LottoNumber>`를 생성하는 방법을 고민해보시면 좋을 것 같아요.

Starlight258 replied:
IntStream을 이용하면 같은 기능을 더 간결하게 작성할 수 있겠네요 !!

jinyoungchoi95 reviewed (approved):
안녕하세요 명지님 😃

피드백 잘 반영해주셨네요. 다음 미션 진행을 위해 머지하겠습니다.

감사합니다🙇‍♂️

Starlight258 committed:
refactor: 프로덕션에 테스트 메서드 제거

- 테스트에서 사용하는 메서드는 테스트에 정의한다.

Starlight258 committed:
refactor: 코드 리팩토링

- Price로 이동한 메서드에 맞춰 테스트도 옮긴다.
- 외부에서 사용하지 않는 메서드는 private으로 변경한다.

Starlight258 committed:
delete: 생성자를 통한 유연한 객체 생성 및 테스트 유틸리티 클래스 제거

- 기존의 복잡한 객체 생성 로직을 생성자로 이동하여 유연하게 객체를 생성할 수 있도록 개선한다.
- 생성자를 테스트 코드에서 활용함으로써 테스트 유틸리티 클래스를 제거하고 실제 프로덕션 코드를 테스트할 수 있다.

Starlight258 committed:
refactor: IntStream과 람다를 이용해 간결하게 리팩토링

Starlight258 committed:
refactor: LottoNumber 캐시 초기화 로직 개선

- 기존에는 LottoNumber를 미리 할당하고 각 원소별로 LottoNumber를 개별적으로 할당했다.
- LottoNumber 배열을 생성한 후 캐시 배열에 할당하도록 수정하여 간결하게 만든다.

Starlight258 commented:
@jinyoungchoi95 
리뷰 반영했습니다! 
세심하게 봐주셔서 감사합니다 🙇‍♀️
